File: benchmarks/suite5/bits.c
#include <stdint.h>
#include <stdio.h>
#include "bench.h"

static int popcount64(int64_t x) {
  int c = 0;
  while (x) {
    c += (int)(x & 1);
    x >>= 1;
  }
  return c;
}

int main(void) {
  const int64_t n = suite5_opaque_i64(25000000);
  int64_t x = 1;
  int64_t acc = 0;
  for (int64_t i = 0; i < n; i++) {
    x = (x * 31 + i) % 1000000007;
    acc += popcount64(x);
  }
  suite5_print_i64(acc);
  return 0;
}

File: benchmarks/suite5/bits.zig
const bench = @import("bench.zig");
const c = @cImport({
    @cInclude("stdio.h");
});

fn popcount64(x: i64) i64 {
    var v = x;
    var cnt: i64 = 0;
    while (v != 0) : (v >>= 1) {
        cnt += v & 1;
    }
    return cnt;
}

pub fn main() void {
    const n: i64 = bench.opaqueI64(25000000);
    var x: i64 = 1;
    var acc: i64 = 0;
    var i: i64 = 0;
    while (i < n) : (i += 1) {
        x = @rem(x * 31 + i, 1000000007);
        acc += popcount64(x);
    }
    bench.printI64(acc);
}
